src/Aspire.Hosting/Dashboard/DcpDataSource.cs

```diff
@@ -0,0 +1,430 @@
+            WorkingDirectory = executable.Spec.WorkingDirectory,
+            Arguments = executable.Spec.Args?.ToImmutableArray(),
+            State = executable.Status?.State,
+            LogSource = new FileLogSource(executable.Status?.StdOutFile, executable.Status?.StdErrFile),
+            ProcessId = executable.Status?.ProcessId,
+            ExpectedEndpointsCount = GetExpectedEndpointsCount(executable),
+            Environment = GetEnvironmentVariables(executable.Status?.EffectiveEnv, executable.Spec.Env),
+            Endpoints = endpoints,
+            Services = services
+        };
+    }
+
+    private (ImmutableArray<string> Endpoints, ImmutableArray<ResourceServiceSnapshot> Services) GetEndpointsAndServices(
+        CustomResource resource,
+        string resourceKind,
+        string? projectPath = null)
+    {
+        var endpoints = ImmutableArray.CreateBuilder<string>();
+        var services = ImmutableArray.CreateBuilder<ResourceServiceSnapshot>();
+        var name = resource.Metadata.Name;
+
+        foreach (var endpoint in _endpointsMap.Values)
+        {
+            if (endpoint.Metadata.OwnerReferences?.Any(or => or.Kind == resource.Kind && or.Name == name) != true)
+            {
+                continue;
+            }
+
+            if (endpoint.Spec.ServiceName is not null
+                && _servicesMap.TryGetValue(endpoint.Spec.ServiceName, out var service)
+                && service?.UsesHttpProtocol(out var uriScheme) == true)
+            {
+                var endpointString = $"{uriScheme}://{endpoint.Spec.Address}:{endpoint.Spec.Port}";
+
+                // For project look into launch profile to append launch url
+                if (projectPath is not null
+                    && _applicationModel.TryGetProjectWithPath(name, projectPath, out var project)
+                    && project.GetEffectiveLaunchProfile() is LaunchProfile launchProfile
+                    && launchProfile.LaunchUrl is string launchUrl)
+                {
+                    if (!launchUrl.Contains("://"))
+                    {
+                        // This is relative URL
+                        endpointString += $"/{launchUrl}";
+                    }
+                    else
+                    {
+                        // For absolute URL we need to update the port value if possible
+                        if (launchProfile.ApplicationUrl is string applicationUrl
+                            && launchUrl.StartsWith(applicationUrl))
+                        {
+                            endpointString = launchUrl.Replace(applicationUrl, endpointString);
+                        }
+                    }
+
+                    // If we cannot process launchUrl then we just show endpoint string
+                }
+
+                endpoints.Add(endpointString);
+            }
+        }
+
+        if (_resourceAssociatedServicesMap.TryGetValue((resourceKind, name), out var resourceServiceMappings))
```

Comment:
I'm trying to remember, is the resourceKind part of this mapping because of the original idea that resource names were only guaranteed unique within a particular resource type? But more recently it's been stated that they'll be globally unique. So maybe we could simplify the key for the mapping now?

And by "now" I don't mean this PR since you've already got a lot of changes here. Just making sure someone else reads this thought 😊

Reply:
I have wondered the same thing. If we go for unique names, we should make a few other changes throughout. It has an impact on #1274 too, as currently we have a composite key (name/kind) in the protocol.